esx_service/utils/vmdk_utils.py

```diff
@@ -63,44 +63,57 @@
 VMDK_RETRY_COUNT = 5
 VMDK_RETRY_SLEEP = 1
 
-def init_datastoreCache():
+# For managing resource locks.
+lockManager = threadutils.LockManager()
+
+
+def init_datastoreCache(force=False):
     """
-    Initializes the datastore cache with the list of datastores accessible from local ESX host.
+    Initializes the datastore cache with the list of datastores accessible
+    from local ESX host. force=True will force it to ignore current cache
+    and force init
     """
     global datastores
     logging.debug("init_datastoreCache: %s", datastores)
 
-    si = vmdk_ops.get_si()
+    with lockManager.get_lock("init_datastoreCache"):
+        if datastores:
+            if force:
+                datastores = None
+            else:
+                return
+
+        si = vmdk_ops.get_si()
 
-    #  We are connected to ESX so childEntity[0] is current DC/Host
-    ds_objects = \
-      si.content.rootFolder.childEntity[0].datastoreFolder.childEntity
-    datastores = []
+        #  We are connected to ESX so childEntity[0] is current DC/Host
+        ds_objects = si.content.rootFolder.childEntity[0].datastoreFolder.childEntity
+        datastores = []
+
+        for datastore in ds_objects:
+            dockvols_path, err = vmdk_ops.get_vol_path(datastore.info.name)
+            if err:
+                continue
+            datastores.append((datastore.info.name,
+                            datastore.info.url,
+                            dockvols_path))
 
-    for datastore in ds_objects:
-        dockvols_path, err = vmdk_ops.get_vol_path(datastore.info.name)
-        if err:
-            continue
-        datastores.append((datastore.info.name,
-                           datastore.info.url,
-                           dockvols_path))
 
 def validate_datastore(datastore):
     """
     Checks if the datastore is part of datastoreCache.
-    If not it will update the datastore cache and checks if datastore is part of the updated cache.
+    If not it will update the datastore cache and check if datastore
+    is a part of the updated cache.
     """
-    global datastores
-    if datastores == None:
-        init_datastoreCache()
+    init_datastoreCache()
     if datastore in [i[0] for i in datastores]:
         return True
     else:
-        init_datastoreCache()
+        init_datastoreCache(force=True)
         if datastore in [i[0] for i in datastores]:
```

Comment:
What happens is one thread is doing this check here (looping thru the datastores while another thread is running init_datastoreCache() and reconstructing the datastores list?